File: .skills/openclaw-skills/skills/hfarazi/ticktickpower/ticktick/commands/batch_abandon.py
import json
import sys

from ..api import api
from ..util import is_task_id


def batch_abandon_command(args) -> None:
    task_ids: list[str] = args.task_ids

    if not task_ids:
        print("Error: At least one task ID is required", file=sys.stderr)
        sys.exit(1)

    invalid = [tid for tid in task_ids if not is_task_id(tid)]
    if invalid:
        print(
            f"Error: Invalid task ID format: {', '.join(invalid)}\n"
            "Task IDs must be 24-character hex strings.",
            file=sys.stderr,
        )
        sys.exit(1)

    try:
        updates: list[dict] = []
        not_found: list[str] = []

        for task_id in task_ids:
            found = api.find_task_by_id(task_id)
            if found:
                updates.append({
                    "id": found["task"]["id"],
                    "projectId": found["projectId"],
                    "status": -1,  # won't do / abandoned
                })
            else:
                not_found.append(task_id)

        if not_found:
            print(f"Warning: Tasks not found: {', '.join(not_found)}", file=sys.stderr)

        if not updates:
            print("Error: No valid tasks to abandon", file=sys.stderr)
            sys.exit(1)

        result = api.batch_tasks({"update": updates})

        if args.json:
            print(json.dumps({
                "abandoned": [u["id"] for u in updates],
                "notFound": not_found,
                "response": result,
            }, indent=2))
            return

        id2error = (result or {}).get("id2error") or {}
        success_count = len(updates) - len(id2error)
        print(f"✓ Abandoned {success_count} task(s)")

        if id2error:
            print("Errors:", file=sys.stderr)
            for tid, err in id2error.items():
                print(f"  {tid}: {err}", file=sys.stderr)

        if not_found:
            print(f"Skipped {len(not_found)} task(s) not found")

    except RuntimeError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

File: .skills/openclaw-skills/skills/halr9000/ticktick-enhanced/scripts/commands/batch-abandon.ts
import { api, UpdateTaskInput } from "../api";

interface BatchAbandonOptions {
  json?: boolean;
}

// Check if string looks like a task ID (24-char hex)
function isTaskId(str: string): boolean {
  return /^[a-f0-9]{24}$/i.test(str);
}

export async function batchAbandonCommand(
  taskIds: string[],
  options: BatchAbandonOptions
): Promise<void> {
  if (taskIds.length === 0) {
    console.error("Error: At least one task ID is required");
    process.exit(1);
  }

  // Validate all inputs are task IDs
  const invalidIds = taskIds.filter((id) => !isTaskId(id));
  if (invalidIds.length > 0) {
    console.error(
      `Error: Invalid task ID format: ${invalidIds.join(", ")}\n` +
        "Task IDs must be 24-character hex strings."
    );
    process.exit(1);
  }

  try {
    // Look up each task to get projectId
    const updates: UpdateTaskInput[] = [];
    const notFound: string[] = [];

    for (const taskId of taskIds) {
      const found = await api.findTaskById(taskId);
      if (found) {
        updates.push({
          id: found.task.id,
          projectId: found.projectId,
          status: -1, // -1 = won't do / abandoned
        });
      } else {
        notFound.push(taskId);
      }
    }

    if (notFound.length > 0) {
      console.error(`Warning: Tasks not found: ${notFound.join(", ")}`);
    }

    if (updates.length === 0) {
      console.error("Error: No valid tasks to abandon");
      process.exit(1);
    }

    // Send batch request
    const result = await api.batchTasks({ update: updates });

    if (options.json) {
      console.log(
        JSON.stringify(
          {
            abandoned: updates.map((u) => u.id),
            notFound,
            response: result,
          },
          null,
          2
        )
      );
      return;
    }

    // Report results
    const successCount = updates.length - (result.id2error ? Object.keys(result.id2error).length : 0);
    console.log(`✓ Abandoned ${successCount} task(s)`);

    if (result.id2error && Object.keys(result.id2error).length > 0) {
      console.error("Errors:");
      for (const [id, error] of Object.entries(result.id2error)) {
        console.error(`  ${id}: ${error}`);
      }
    }

    if (notFound.length > 0) {
      console.log(`Skipped ${notFound.length} task(s) not found`);
    }
  } catch (error) {
    console.error(
      `Error: ${error instanceof Error ? error.message : String(error)}`
    );
    process.exit(1);
  }
}
